kotest-property/kotest-property-core/src/commonMain/kotlin/io/kotest/property/core/proptest.kt

```diff
@@ -0,0 +1,59 @@
+package io.kotest.property.core
+
+import io.kotest.mpp.sysprop
+import io.kotest.property.AssumptionFailedException
+import io.kotest.property.Gen
+import io.kotest.property.PropertyTesting
+import io.kotest.property.RandomSource
+import io.kotest.property.ShrinkingMode
+import kotlin.random.Random
+import kotlin.time.Duration
+
+class PropTestContext {
+
+   var iterations: Int? = null // converted to iteration based constraint
+   var duration: Duration? = null // converted to duration based constraint
+   var constraints: Constraints? = null // specify manual constraits
+   var minSuccess = -1
+   var maxFailure = -1
+   var edgecaseFactor = PropertyTesting.defaultEdgecasesGenerationProbability
+   var shouldPrintShrinkSteps = false // sysprop("kotest.proptest.output.shrink-steps", true)
+   var shrinkingMode: ShrinkingMode = PropertyTesting.defaultShrinkingMode
+   var maxDiscardPercentage: Int = 20
+   var shouldPrintGeneratedValues: Boolean = PropertyTesting.shouldPrintGeneratedValues
+
+   // must call seed before rs otherwise has no effect
+   var seed: Long? = null
+   var failOnSeed: Boolean = sysprop("kotest.proptest.seed.fail-if-set", false)
```

Comment:
We do something similar with defaultTestConfig but also having all the top level fields.
I guess one could wrap all this in a config {} block so its contained. From what I can see, that seems to be how other kotlin frameworks do config, rather than creating a data class and passing it in.

Reply:
Late to the party here.. Do you have some examples of other Kotlin libs/frameworks using this approach? I know we started doing it a bit for some assertions, but I'm starting to have second doubts about whether that was actually a good call.

There's a few clear upsides to having it as function params instead of mutable context.

```
      proptest("shrinking").config(
         seed = 838127382173, // override seed
         shouldPrintShrinkSteps = true, // print shrink steps
      ) {
         val a by gen { Arb.int() }
         val b by gen { Arb.int() }

         property {
            a * b shouldBe a + b
         }
      }
```